keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,499 @@
+
+Delegating EndpointSlice control would allow custom controllers to define their own criteria for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As a reference implementation, and since [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.
+
+Providing a generic EndpointSlice Reconciler module would allow users to reuse the same code as the Kubernetes EndpointSlice Controller, thus ensuring consistency and reducing the effort needed to implement custom reconciliation logic. As highlighted in this [comment](https://github.com/kubernetes/kubernetes/pull/118953#discussion_r1245970845), the EndpointSlice Mirroring Controller would also benefit from these changes since the generic EndpointSlice Reconciler could be used from the EndpointSlice Mirroring Controller.
+
+### Goals
+
+* Provide the ability to disable the Kubernetes EndpointSlice controller for particular services.
+* Extend and enhance the EndpointSlice Reconciler module to make it generic.
+
+### Non-Goals
+
+* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice controller.
+* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
+* Modify the Service / EndpointSlice Specs.
+* Add new features to the EndpointSlice Mirroring Controller.
+
+## Proposal
+
+#### Well-Known Label
+
+`service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, no matter the service specs, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers for that service will be disabled, thus Endpoints and EndpointSlices for this service will not be created by the Kubernetes Controller Manager. If the label is not set, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers will be enabled for that service and the Endpoints and EndpointSlices will be handled as of today.
+
+The Kubernetes Controller Manager will implement this label both at object creation and on dynamic addition/removal/updates of this label.
+
+#### Flexibility on the EndpointSlice Reconciler Module
+
+The reconciler structure will support more features to cover the requirements set by the current behavior of the EndpointSlice Mirroring Controller (e.g.: placeholder EndpointSlice does not exist in the Mirroring Controller).
+
+The `Reconcile` function definition will be changed to accept general types of data (not specific to Services) and the list of features the endpointslices being reconciled will support (e.g. Traffic Distribution). Functions for specific types (Service/Pods for the EndpointSlice Controller and Endpoints for the EndpointSlice Mirroring Controller) generating data to pass to the `Reconcile` function will be provided by the EndpointSlice Reconciler module.
```

Comment:
This sounds particularly risky, we'll need to tread very carefully here when changing the signature of such a fundamental function. Worse, much of our test coverage is built around this function, so it would be hard to keep the same tests through this change.

Reply:
The unit tests are not complex to modify, a helper function is used (https://github.com/kubernetes/endpointslice/blob/v0.30.3/reconciler_test.go#L2313). I changed it in the first version I made: https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3437. This is for Pod/Service objects.

To reconcile Endpoints, we could just use another helper function: https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3677.